### PR TITLE
txpool refactor: support multiple implementations via trait object

### PR DESCRIPTION
**Pull request by michalkucharczyk**
This pull request slightly refactors the transaction pool crate, laying the groundwork for a fork-aware implementation (which will be provided as separate PR). The main goal of this refactor is to allow  multiple implementations of transaction-pool. This PR does not change any logic yet, it is intended to make the review of the fork-aware transaction pool PR easier.

Key changes include:

1. **Transition to `TransactionPoolImpl` Trait Object**:
   - Replaced `sc_transaction_pool::FullPool` with `sc_transaction_pool::TransactionPoolImpl`.
   - Updated references and type definitions across multiple files.

2. **Builder Pattern**:
   - Introduced `TransactionPoolType` enum for different implementations.
   - Created a `Builder` struct for constructing transaction pool trait object.
   - Enhanced `TransactionPoolOptions` for parameter-based initialization.

3. **Modularization and Refactoring**:
   - Moved components that will be re-used by fork-aware implementation (e.g., error handling, enactment state, metrics) to a `common` module.
   - Created a `single_state_txpool` module for the existing implementation of transaction pool.
   - Renamed and moved several files for clarity, adjusting import paths accordingly.

4. **Support for `?Sized` Trait**:
   - Updated struct definitions and methods to support the `?Sized` trait for `TransactionPool`.
   - Applied these updates across various files.

Related to: #1202

### CODE REVIEW

**Review by davxy — COMMENTED** [review]
I see that `?Sized` bound is spread all over the place.
What about adding the bound directly to the definition of the `TransactionPool` and `MaintainedTransactionPool` traits?

File: polkadot/rpc/src/lib.rs [hunk 1]
```diff
@@ -112,7 +112,7 @@ where
 	C::Api: pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>,
 	C::Api: BabeApi<Block>,
 	C::Api: BlockBuilder<Block>,
-	P: TransactionPool + Sync + Send + 'static,
+	P: TransactionPool + Sync + Send + 'static + ?Sized,
```
Comment (davxy) [inline polkadot/rpc/src/lib.rs:115]:
```suggestion
	P: TransactionPool + 'static + ?Sized,
```
Nit. Already bound by Send+Sync. Probably there are other instance around.

File: substrate/client/transaction-pool/src/builder.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+			Error = <FullChainApi<Client, Block> as ChainApi>::Error,
+		>,
+{
+}
+
+/// The type alias for a dynamic trait object implementing
+/// `FullClientTransactionPool` with the given `Client` and `Block` types.
+///
+/// This trait object abstracts away the specific implementations of the transaction pool.
+pub type TransactionPoolImpl<Block, Client> = dyn FullClientTransactionPool<Block, Client>;
```
Comment (davxy) [inline substrate/client/transaction-pool/src/builder.rs:165]: Nit. I'm not sure about this name, I'm referring to the `Impl` suffix. But I can't suggest a better name :-)

File: templates/parachain/node/src/rpc.rs [hunk 3]
```diff
@@ -42,7 +42,7 @@ where
 	C::Api: pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>,
 	C::Api: substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>,
 	C::Api: BlockBuilder<Block>,
-	P: TransactionPool + Sync + Send + 'static,
+	P: TransactionPool + Sync + Send + 'static + ?Sized,
```
Comment (davxy) [inline templates/parachain/node/src/rpc.rs:45]:
```suggestion
	P: TransactionPool +  'static + ?Sized,
```

File: substrate/client/transaction-pool/src/builder.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+/// Builder allowing to create specific instance of transaction pool.
+pub struct Builder<Block, Client> {
+	options: TransactionPoolOptions,
+	_phantom: PhantomData<(Client, Block)>,
+}
```
Comment (bkchr) [inline substrate/client/transaction-pool/src/builder.rs:171]: This is not really a builder pattern. Not really sure we need this.

File: substrate/client/service/src/builder.rs [hunk 5]
```diff
@@ -349,7 +349,7 @@ where
 }
 
 /// Parameters to pass into `build`.
-pub struct SpawnTasksParams<'a, TBl: BlockT, TCl, TExPool, TRpc, Backend> {
+pub struct SpawnTasksParams<'a, TBl: BlockT, TCl, TExPool: ?Sized, TRpc, Backend> {
```
Comment (bkchr) [inline substrate/client/service/src/builder.rs:352]: All these `?Sized` are "confusing". IMO we should just make `transaction_pool: Arc<dyn TransactionPool>` and thats it.

File: substrate/client/transaction-pool/src/builder.rs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+};
+use prometheus_endpoint::Registry as PrometheusRegistry;
+use sc_transaction_pool_api::{LocalTransactionPool, MaintainedTransactionPool};
+use sp_core::traits::SpawnEssentialNamed;
+
+/// The type of transaction pool.
+#[derive(Debug, Clone)]
+pub enum TransactionPoolType {
+	/// Single-state transaction pool
+	SingleState,
```
Comment (bkchr) [inline substrate/client/transaction-pool/src/builder.rs:35]:
```suggestion
	NonForkAware,
```

Maybe this way? Not sure, but `SingleState` is weird. If you keep the name, at least be a little bit more descriptive.

File: substrate/client/cli/src/params/transaction_pool_params.rs [hunk 7]
```diff
@@ -16,8 +16,25 @@
...
-use sc_service::config::TransactionPoolOptions;
+use clap::{Args, ValueEnum};
+use sc_transaction_pool::TransactionPoolOptions;
+
+/// Type of transaction pool to be used
+#[derive(Debug, Clone, Copy, ValueEnum)]
+#[value(rename_all = "kebab-case")]
+pub enum TransactionPoolType {
+	/// Uses a legacy, single-state transaction pool.
+	SingleState,
```
Comment (bkchr) [inline substrate/client/cli/src/params/transaction_pool_params.rs:27]: Yeah or just call it `LegacyTxPool`.

**Review by bkchr — COMMENTED** [review]
I think the approach with making every unsized is not really sexy. If we touch all of this stuff any way, we should directly put transaction pool everywhere as `Arc<dyn TransactionPool>` or whatever.

File: substrate/client/transaction-pool/src/builder.rs (hunk 4 above)
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/builder.rs:171]: We need to translate options to instance. I can rename it, but we need some entry point to instantiate.

File: substrate/client/service/src/builder.rs (hunk 5 above)
Comment (michalkucharczyk) [reply substrate/client/service/src/builder.rs:352]: Do I understand correctly that you want to remove `TransactionPool` generic type parameter from all generic types?

Comment (bkchr) [reply substrate/client/service/src/builder.rs:352]: Yes, would probably make it easier.

**Pull request closed without merging**